File: desktop_app/gemini_client.py
"""Gemini Vision client for answering questions about a screenshot.

Uses the ``google-genai`` SDK if available (preferred), otherwise falls
back to a direct REST call against Google's Generative Language API.
This keeps the install small for users who want a quick start.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Optional

import httpx


@dataclass
class GeminiResult:
    text: str
    raw: dict
    model: str


class GeminiClient:
    """Minimal Gemini Vision client built on top of httpx."""

    DEFAULT_ENDPOINT = (
        "https://generativelanguage.googleapis.com/v1beta/models/"
        "{model}:generateContent"
    )

    def __init__(
        self,
        api_key: str,
        model: str = "gemini-2.0-flash",
        timeout: float = 60.0,
    ) -> None:
        if not api_key:
            raise ValueError("Gemini API key is required.")
        self.api_key = api_key
        self.model = model
        self._client = httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "GeminiClient":
        return self

    def __exit__(self, *_exc) -> None:
        self.close()

    def answer_about_image(
        self,
        png_bytes: bytes,
        question: str,
        mime_type: str = "image/png",
        system_instruction: Optional[str] = None,
    ) -> GeminiResult:
        """Send a question + image to Gemini and return the answer text."""
        b64 = base64.standard_b64encode(png_bytes).decode("ascii")
        payload: dict = {
            "contents": [
                {
                    "role": "user",
                    "parts": [
                        {"text": question},
                        {
                            "inline_data": {
                                "mime_type": mime_type,
                                "data": b64,
                            }
                        },
                    ],
                }
            ],
            "generationConfig": {
                "temperature": 0.4,
                "maxOutputTokens": 2048,
            },
        }
        if system_instruction:
            payload["systemInstruction"] = {
                "role": "system",
                "parts": [{"text": system_instruction}],
            }

        url = self.DEFAULT_ENDPOINT.format(model=self.model)
        response = self._client.post(
            url,
            params={"key": self.api_key},
            headers={"Content-Type": "application/json"},
            content=json.dumps(payload),
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"Gemini API error {response.status_code}: {response.text[:500]}"
            )
        body = response.json()
        text = _extract_text(body)
        return GeminiResult(text=text, raw=body, model=self.model)


def _extract_text(body: dict) -> str:
    """Pull the assistant text out of a Gemini ``generateContent`` response."""
    candidates = body.get("candidates") or []
    if not candidates:
        prompt_feedback = body.get("promptFeedback")
        if prompt_feedback:
            return f"(Gemini blocked the prompt: {prompt_feedback})"
        return "(no response)"
    parts = candidates[0].get("content", {}).get("parts") or []
    chunks = [p.get("text", "") for p in parts if isinstance(p, dict)]
    return "".join(chunks).strip() or "(empty response)"

File: android/app/src/main/java/com/draura/aura/gemini/GeminiClient.kt
package com.draura.aura.gemini

import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Minimal Gemini REST client using the public Generative Language API.
 *
 * This deliberately avoids the official Google AI client SDK so the app
 * stays small and we don't pull in a transitive dep on Guava + Protobuf.
 */
class GeminiClient {
    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    /**
     * Sends [pngBytes] (the screenshot) plus [question] to Gemini and
     * returns the model's plain-text answer.
     *
     * @throws GeminiException on HTTP or API errors.
     */
    suspend fun answer(
        apiKey: String,
        model: String,
        pngBytes: ByteArray,
        question: String,
    ): String = withContext(Dispatchers.IO) {
        require(apiKey.isNotBlank()) { "Gemini API key is required" }

        val b64 = Base64.encodeToString(pngBytes, Base64.NO_WRAP)
        val payload = JSONObject().apply {
            put("contents", JSONArray().put(
                JSONObject().put("parts", JSONArray()
                    .put(JSONObject().put("text", question))
                    .put(JSONObject().put(
                        "inline_data",
                        JSONObject()
                            .put("mime_type", "image/png")
                            .put("data", b64)
                    ))
                )
            ))
            put("systemInstruction", JSONObject().put("parts", JSONArray().put(
                JSONObject().put(
                    "text",
                    "You are an expert tutor helping the user answer the question " +
                        "in the attached screenshot. Be precise. If multiple choice, " +
                        "give the answer letter and a 1-2 line justification."
                )
            )))
            put("generationConfig", JSONObject()
                .put("temperature", 0.2)
                .put("maxOutputTokens", 1024))
        }

        val url = "https://generativelanguage.googleapis.com/v1beta/models/" +
            "$model:generateContent?key=$apiKey"
        val req = Request.Builder()
            .url(url)
            .post(payload.toString().toRequestBody(JSON))
            .build()

        http.newCall(req).execute().use { resp ->
            val body = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) {
                throw GeminiException(
                    "Gemini HTTP ${resp.code}: ${truncate(body, 400)}"
                )
            }
            extractText(body)
        }
    }

    private fun extractText(body: String): String {
        val root = JSONObject(body)
        val candidates = root.optJSONArray("candidates")
            ?: throw GeminiException("Empty response from Gemini.")
        if (candidates.length() == 0) {
            val pf = root.optJSONObject("promptFeedback")
            throw GeminiException(
                "Gemini returned no candidates. ${pf?.toString()?.let { "Reason: $it" }.orEmpty()}"
            )
        }
        val parts = candidates.getJSONObject(0)
            .optJSONObject("content")
            ?.optJSONArray("parts")
            ?: throw GeminiException("Malformed Gemini response.")
        val sb = StringBuilder()
        for (i in 0 until parts.length()) {
            val t = parts.getJSONObject(i).optString("text", "")
            if (t.isNotEmpty()) sb.append(t)
        }
        return sb.toString().ifBlank { "(empty response)" }
    }

    private fun truncate(s: String, max: Int): String =
        if (s.length <= max) s else s.substring(0, max) + "…"

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()
    }
}

class GeminiException(msg: String) : RuntimeException(msg)
